src/Streamly/External/ByteString.hs

```diff
@@ -14,38 +14,67 @@ where
-import Streamly.Data.Unfold (Unfold, lmap)
+import Foreign.ForeignPtr (withForeignPtr)
+import Foreign.Storable (peek)
+import GHC.Exts
+    ( Addr#
+    , MutableByteArray#
+    , RealWorld
+    , byteArrayContents#
+    , minusAddr#
+    , plusAddr#
+    , unsafeCoerce#
+    )
+import GHC.ForeignPtr (ForeignPtr(..), ForeignPtrContents(..))
+import GHC.Int (Int(..))
+import GHC.Ptr (Ptr(..), nullPtr, plusPtr)
 import Streamly.Data.Fold (Fold)
+import Streamly.Data.Unfold (Unfold, lmap)
 
 -- Internal imports
 import Data.ByteString.Internal (ByteString(..))
-import Streamly.Internal.Data.Array.Foreign.Type (Array(..))
-import Streamly.Internal.Data.Array.Foreign.Mut.Type
-    (ArrayContents, arrayToFptrContents, fptrToArrayContents, nilArrayContents)
+import Streamly.Internal.Data.Array.Unboxed.Mut.Type (nilArrayContents)
+import Streamly.Internal.Data.Array.Unboxed.Type (Array(..))
+import Streamly.Data.Unbox (MutableByteArray(..))
+import Streamly.Internal.System.IO (unsafeInlineIO)
 
-import qualified Streamly.Data.Array.Foreign as A
+import qualified Streamly.Data.Array.Unboxed as Array
+import qualified Streamly.Internal.Data.Unfold as Unfold
+import qualified Streamly.Internal.Data.Stream.StreamD as StreamD
 
 import Prelude hiding (read)
 
+{-# INLINE mutableByteArrayContents# #-}
+mutableByteArrayContents# :: MutableByteArray# RealWorld -> Addr#
+mutableByteArrayContents# marr# = byteArrayContents# (unsafeCoerce# marr#)
+
 -- | Helper function that creates a ForeignPtr
-makeForeignPtr :: ArrayContents -> Ptr a -> ForeignPtr a
-makeForeignPtr contents (Ptr addr#) =
-    ForeignPtr addr# (arrayToFptrContents contents)
+{-# INLINE makeForeignPtr #-}
+makeForeignPtr :: MutableByteArray a -> Int -> ForeignPtr a
+makeForeignPtr (MutableByteArray marr#) (I# off#) =
+    ForeignPtr
+        (mutableByteArrayContents# marr# `plusAddr#` off#)
+        (PlainPtr marr#)
 
 -- | Convert a 'ByteString' to an array of 'Word8'. This function unwraps the
 -- 'ByteString' and wraps it with 'Array' constructors and hence the operation
 -- is performed in constant time.
 {-# INLINE toArray #-}
 toArray :: ByteString -> Array Word8
-toArray (PS (ForeignPtr addr# _) _ _)
-    | Ptr addr# == nullPtr = Array nilArrayContents nullPtr nullPtr
-toArray (PS (ForeignPtr addr# fpcontents) off len) =
-    Array (fptrToArrayContents fpcontents) startPtr endPtr
-  where
-    startPtr = Ptr addr# `plusPtr` off
-    endPtr = startPtr `plusPtr` len
+toArray (BS (ForeignPtr addr# _) _)
+    | Ptr addr# == nullPtr = Array nilArrayContents 0 0
+toArray (BS (ForeignPtr addr# (PlainPtr marr#)) len) =
+    let off = I# (addr# `minusAddr#` mutableByteArrayContents# marr#)
+     in Array (MutableByteArray marr#) off (off + len)
+toArray (BS fptr len) =
+    unsafeInlineIO
+        $ withForeignPtr fptr $ Unfold.fold (Array.writeN len) generator
+
+    where
+
+    generator =
+        Unfold.mkUnfoldrM
+            (\ptr -> flip StreamD.Yield (ptr `plusPtr` 1) <$> peek ptr)
```

Comment:
We can use `fromPtr` here maybe? That will result in a pinned mutable byte array.